modyn/trainer_server/internal/trainer/remote_downsamplers/remote_rho_loss_downsampling.py

```diff
@@ -0,0 +1,69 @@
+from typing import Optional, Union
+
+import torch
+from modyn.trainer_server.internal.trainer.remote_downsamplers.abstract_remote_downsampling_strategy import (
+    AbstractRemoteDownsamplingStrategy,
+)
+from modyn.trainer_server.internal.trainer.remote_downsamplers.irreducible_loss_producer import IrreducibleLossProducer
+
+
+class RemoteRHOLossDownsampling(AbstractRemoteDownsamplingStrategy):
+    """
+    Method adapted from Prioritized Training on Points that are Learnable, Worth Learning, and Not Yet Learnt
+    (Sören Mindermann+, 2022). https://arxiv.org/abs/2206.07137
+    """
+
+    def __init__(
+        self,
+        pipeline_id: int,
+        trigger_id: int,
+        batch_size: int,
+        params_from_selector: dict,
+        modyn_config: dict,
+        per_sample_loss: torch.nn.modules.loss,
+        device: str,
+    ) -> None:
+        super().__init__(pipeline_id, trigger_id, batch_size, params_from_selector, modyn_config, device)
+        self.per_sample_loss_fct = per_sample_loss
+        self.irreducible_loss_producer = IrreducibleLossProducer(
+            per_sample_loss,
+            modyn_config,
+            params_from_selector["rho_pipeline_id"],
+            params_from_selector["il_model_id"],
+            device,
+        )
+        self.rho_loss: torch.Tensor = torch.tensor([])
+        self.number_of_points_seen = 0
+
+    def init_downsampler(self) -> None:
+        self.index_sampleid_map: list[int] = []
+        self.rho_loss = torch.tensor([])
+        self.number_of_points_seen = 0
+
+    def inform_samples(
+        self,
+        sample_ids: list[int],
+        forward_input: Union[dict[str, torch.Tensor], torch.Tensor],
+        forward_output: torch.Tensor,
+        target: torch.Tensor,
+        embedding: Optional[torch.Tensor] = None,
+    ) -> None:
+        training_loss = self.per_sample_loss_fct(forward_output, target).detach()
+        self.index_sampleid_map += sample_ids
+        irreducible_loss = self.irreducible_loss_producer.get_irreducible_loss(sample_ids, forward_input, target)
+        self.rho_loss = torch.cat([self.rho_loss, training_loss - irreducible_loss]).to(training_loss.dtype)
```

Comment:
Why is the type conversation here necessary? Shouldn't all loss tensors be of the same dtype?

Reply:
Note we initialize `self.rho_loss` as `torch.tensor([])`. This empty tensor has dtype float32 (pytorch default dtype) irregardless of the loss type.